clang/tools/libclang/CDependencies.cpp

```diff
@@ -264,6 +280,53 @@ CXErrorCode clang_experimental_DependencyScannerWorker_getFileDependencies_v4(
         }
         *Out = new CXFileDependenciesList{FD.Commands.size(), Commands};
       });
+
+  return Result;
+}
+
+CXErrorCode clang_experimental_DependencyScannerWorker_getFileDependencies_v5(
+    CXDependencyScannerWorker W, int argc, const char *const *argv,
+    const char *ModuleName, const char *WorkingDirectory, void *MDCContext,
+    CXModuleDiscoveredCallback *MDC, void *MLOContext,
+    CXModuleLookupOutputCallback *MLO, unsigned, CXFileDependenciesList **Out,
+    CXDiagnosticSet *OutDiags) {
+  OutputLookup OL(MLOContext, MLO);
+  auto LookupOutputs = [&](const ModuleID &ID, ModuleOutputKind MOK) {
+    return OL.lookupModuleOutput(ID, MOK);
+  };
+
+  if (!Out)
+    return CXError_InvalidArguments;
+  *Out = nullptr;
+
+  CXDiagnosticSetDiagnosticConsumer DiagConsumer;
+
+  CXErrorCode Result = getFileDependencies(
+      W, argc, argv, WorkingDirectory, MDC, MDCContext, nullptr, &DiagConsumer,
+      LookupOutputs,
+      /*DeprecatedDriverCommand=*/false,
+      ModuleName ? Optional<StringRef>(ModuleName) : None,
+      [&](FullDependencies FD) {
+        assert(FD.DriverCommandLine.empty());
+        std::vector<std::string> Modules;
+        for (const ModuleID &MID : FD.ClangModuleDeps)
+          Modules.push_back(MID.ModuleName + ":" + MID.ContextHash);
+        auto *Commands = new CXTranslationUnitCommand[FD.Commands.size()];
+        for (size_t I = 0, E = FD.Commands.size(); I < E; ++I) {
+          Commands[I].ContextHash = cxstring::createDup(FD.ID.ContextHash);
+          Commands[I].FileDeps = cxstring::createSet(FD.FileDeps);
+          Commands[I].ModuleDeps = cxstring::createSet(Modules);
+          Commands[I].Executable =
+              cxstring::createDup(FD.Commands[I].Executable);
+          Commands[I].BuildArguments =
+              cxstring::createSet(FD.Commands[I].Arguments);
+        }
+        *Out = new CXFileDependenciesList{FD.Commands.size(), Commands};
+      });
+
+  *OutDiags = DiagConsumer.getDiagnosticSet();
```

Comment:
Just making sure it's intentional to allocate `OutDiags` even on success.  I think this is fine -- right now it's only emitting errors, but I could imagine including warnings in the future, and we don't want to change whether this out parameter is set or not.

Reply:
Agree, I think the API is simpler this way.